### PR TITLE
KAFKA-15838: KIP-1040: Add replace.null.with.default to InsertField, ExtractField, HeaderFrom, Cast, SetSchemaMetadata, TimestampConverter, MaskField transforms

### PR DESCRIPTION
**Pull request by mfvitale**
*Value `null` is valid for an optional filed, even though the field has a default value.
Only when field is required, the class return default value fallback when value is `null`*

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mfvitale** [comment]
A follow up after this https://issues.apache.org/jira/browse/KAFKA-8713

@C0urante @divijvaidya @mimaison

**Comment by mimaison** [comment]
Hi @mfvitale, thanks for the PR!
This is adding new configurations to transformations so this will require a [KIP](https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Improvement+Proposals).

**Comment by mfvitale** [comment]
> Hi @mfvitale, thanks for the PR! This is adding new configurations to transformations so this will require a [KIP](https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Improvement+Proposals).

@mimaison actually `key.converter.replace.null.with.default` and `value.converter.replace.null.with.default` are not new configurations but are the one used by `JsonConverter` added with https://github.com/apache/kafka/pull/13419 to fix https://issues.apache.org/jira/browse/KAFKA-8713, Isn't it?

**Comment by mimaison** [comment]
While it's not directly adding new configurations, it's effectively changing the behavior of `ExtractField` and `InsertField` and making them support new configurations. The `configure()` method on Transformation is only called with configurations provided with their prefix. In this example a user would need to set something like:

```
transforms=my-smt
transforms.my-smt.type=MyTransformation
transforms.my-smt.key.converter.replace.null.with.default=false
```

Also by not defining the configurations explicitly users can't discover them using the `GET /connector-plugins/{plugin-type}/config`.

**Comment by mfvitale** [comment]
> The `configure()` method on Transformation is only called with configurations provided with their prefix.

This cleared my doubt. So in that case it should be clear declared as an SMT configuration. I'll open a KIP. Thanks for the explanation.

**Comment by mfvitale** [comment]
@mimaison Seems that the [KIP-1040](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1040%3A+Improve+handling+of+nullable+values+in+InsertField%2C+ExtractField%2C+and+other+transformations) has been [accepted](https://lists.apache.org/list.html?dev@kafka.apache.org). Do you think it's mergiable now??

**Comment by mfvitale** [comment]
@jlprat do you think this can be backported? It will be beneficial also for Debezium users.

**Comment by mimaison** [comment]
@mfvitale Merge-able, I don't know yet, but we can start reviewing it :)

This is not a candidate for 3.8 as we're now past code freeze. Also the KIP only got 3 binding votes earlier this week and KIP freeze was on May 15.

**Comment by mfvitale** [comment]
> @mfvitale Merge-able, I don't know yet, but we can start reviewing it :)

;-)

> This is not a candidate for 3.8 as we're now past code freeze. Also the KIP only got 3 binding votes earlier this week and KIP freeze was on May 15.

Ok, I tried :smile:

**Comment by mfvitale** [comment]
@C0urante Can I ask to have a look?

**Comment by Mon-MFEC** [comment]
As @mfvitale mentioned.
@divijvaidya @mimaison

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/ExtractField.java [hunk 1]
```diff
@@ -41,21 +41,28 @@ public abstract class ExtractField<R extends ConnectRecord<R>> implements Transf
...
+                            ConfigDef.Type.STRING,
+                            ConfigDef.NO_DEFAULT_VALUE,
+                            ConfigDef.Importance.MEDIUM,
+                            "Field name to extract."
+                    )
+                    .define(REPLACE_NULL_WITH_DEFAULT_CONFIG,
+                            ConfigDef.Type.BOOLEAN,
+                            Boolean.TRUE,
+                            ConfigDef.Importance.MEDIUM,
+                            "Determine if null field value must be replaced with its default value or not."));
```
Comment (mimaison) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/ExtractField.java]: Can we use the same message as in https://github.com/apache/kafka/blob/trunk/connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverterConfig.java#L51?

Same in the other file below.

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/ExtractField.java [hunk 2]
```diff
@@ -41,21 +41,28 @@ public abstract class ExtractField<R extends ConnectRecord<R>> implements Transf
...
+                    .define(
+                            FIELD_CONFIG,
+                            ConfigDef.Type.STRING,
+                            ConfigDef.NO_DEFAULT_VALUE,
+                            ConfigDef.Importance.MEDIUM,
+                            "Field name to extract."
+                    )
+                    .define(REPLACE_NULL_WITH_DEFAULT_CONFIG,
+                            ConfigDef.Type.BOOLEAN,
+                            Boolean.TRUE,
```
Comment (mimaison) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/ExtractField.java]: -> `true`
Same in the other file below.

File: connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java [hunk 3]
```diff
@@ -186,7 +189,98 @@ public void nonExistentNestedFieldWithSchemaShouldFail() {
 
     @Test
     public void testExtractFieldVersionRetrievedFromAppInfoParser() {
-        assertEquals(AppInfoParser.getVersion(), xform.version());
+        assertEquals(AppInfoParser.getVersion(), xformKey.version());
+    }
+
+    @Test
+    @DisplayName("Null key value instead of default must be used")
```
Comment (mimaison) [inline connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java]: We don't set display names on tests in Connect.

File: connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java [hunk 4]
```diff
@@ -186,7 +189,98 @@ public void nonExistentNestedFieldWithSchemaShouldFail() {
...
+    public void whenKeyConverterReplaceNullWithDefaultIsFalseExtractFieldWithNullValueMustReturnNull() {
```
Comment (mimaison) [inline connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java]: Can we simplify the test name? What about `testUnsetOptionalFieldIsNull()`?
Same in the other file below.

File: connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java [hunk 5]
```diff
@@ -186,7 +189,98 @@ public void nonExistentNestedFieldWithSchemaShouldFail() {
...
+
+        final SinkRecord transformedRecord = xformKey.apply(record);
+        Integer extractedValue = (Integer) transformedRecord.key();
+
+        assertNull(extractedValue);
+    }
+
+    @Test
+    @DisplayName("Default key value must be used")
+    public void whenKeyConverterReplaceNullWithDefaultIsTrueExtractFieldWithNullValueMustReturnDefaultValueWhenPresent() {
```
Comment (mimaison) [inline connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java]: Can we simplify the test name? What about `testUnsetOptionalFieldReplacedWithDefault()`?
Same in the other file below.

**Review by mimaison — COMMENTED** [review]
Thanks for the PR.
The KIP mentions updating `InsertField, ExtractField, HeaderFrom, Cast, SetSchemaMetadata, TimestampConverter, MaskField, ValueToKey and ReplaceField`. Is there a reason this PR only updates ExtractField and InsertField?

**Comment by mfvitale** [comment]
@mimaison , thanks for your review. The required changes is now addressed. 

> Thanks for the PR. The KIP mentions updating `InsertField, ExtractField, HeaderFrom, Cast, SetSchemaMetadata, TimestampConverter, MaskField, ValueToKey and ReplaceField`. Is there a reason this PR only updates ExtractField and InsertField?

The orignal KIP scope was for InsertField and ExtractField but it was enlarged during KIP discussion. In agreed with @C0urante that not all SMTs should have implemented in this PR, mainly for time constrain. 

In any case I can try to do my best to add the others, but then will be good if we can merge what we will have ready for the 3.9 feature/code freeze. WDYT?

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java [hunk 6]
```diff
@@ -111,6 +117,7 @@ public String toString() {
     private Map<String, Schema.Type> casts;
     private Schema.Type wholeValueCastType;
     private Cache<Schema, Schema> schemaUpdateCache;
+    private Boolean replaceNullWithDefault;
```
Comment (mimaison) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java]: We can use the primitive type here

**Comment by mimaison** [comment]
Sure, it's fine to do the remaining transformations in another PR. I just wanted to check you were happy with the content of this PR before spending time reviewing it.

**Comment by mfvitale** [comment]
> Sure, it's fine to do the remaining transformations in another PR. I just wanted to check you were happy with the content of this PR before spending time reviewing it.

Yes, I am good with it. I'll let the others SMTs to another PR.

File: connect/transforms/src/test/java/org/apache/kafka/connect/transforms/SetSchemaMetadataTest.java [hunk 7]
```diff
@@ -104,6 +115,38 @@ public void schemaNameAndVersionUpdateWithStruct() {
...
 
+    @ParameterizedTest
+    @MethodSource("data")
+    public void schemaNameAndVersionUpdateWithStructAndNullField(boolean replaceNullWithDefault, Object expectedValue) {
+        final String fieldName1 = "f1";
+        final String fieldName2 = "f2";
+        final int fieldValue2 = 1;
+        final Schema schema = SchemaBuilder.struct()
+                .name("my.orig.SchemaDefn")
+                .field(fieldName1, SchemaBuilder.string().defaultValue("default").optional().build())
```
Comment (mimaison) [inline connect/transforms/src/test/java/org/apache/kafka/connect/transforms/SetSchemaMetadataTest.java:126]: Should we use `expectedValue` instead of `default`?

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java [hunk 8]
```diff
@@ -104,6 +107,8 @@ public static InsertionSpec parse(String spec) {
     private InsertionSpec staticField;
     private String staticValue;
 
+    protected boolean replaceNullWithDefault;
```
Comment (mimaison) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java]: Why protected? Also let's remove the extra empty line just above

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/HeaderFrom.java [hunk 9]
```diff
@@ -105,6 +109,8 @@ public String toString() {
 
     private Operation operation;
 
+    private Boolean replaceNullWithDefault;
```
Comment (mimaison) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/HeaderFrom.java]: `Boolean` -> `boolean`

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/HeaderFrom.java [hunk 10]
```diff
@@ -70,15 +71,17 @@ NO_DEFAULT_VALUE, new NonEmptyListValidator(),
...
+                    "Whether to replace fields that have a default value and that are null to the default value. When set to true, the default value is used, otherwise null is used.");
 
     private final Cache<Schema, Schema> moveSchemaCache = new SynchronizedCache<>(new LRUCache<>(16));
     enum Operation {
         MOVE(MOVE_OPERATION),
         COPY(COPY_OPERATION);
 
-        private final String name;
 
+        private final String name;
```
Comment (mimaison) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/HeaderFrom.java]: Can we undo this change?

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java [hunk 11]
```diff
@@ -54,6 +54,7 @@ private interface ConfigName {
         String TIMESTAMP_FIELD = "timestamp.field";
         String STATIC_FIELD = "static.field";
         String STATIC_VALUE = "static.value";
+        String REPLACE_NULL_WITH_DEFAULT_CONFIG = "replace.null.with.default";
```
Comment (mimaison) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java]: None of the other field have the `_CONFIG` suffix, so let's remove it so we're consistent.
Same in other files.

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/HeaderFrom.java [hunk 12]
```diff
@@ -178,6 +184,14 @@ private R applySchemaless(R record, Object operatingValue) {
         return newRecord(record, null, updatedValue, updatedHeaders);
     }
 
+    private Object getFieldValue(Struct value, Field field) {
+
```
Comment (mimaison) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/HeaderFrom.java]: I think we can remove this empty line.
Same in other files.

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java [hunk 13]
```diff
@@ -243,6 +257,7 @@ public ConfigDef config() {
 
     protected abstract R newRecord(R record, Schema updatedSchema, Object updatedValue);
 
+
```
Comment (mimaison) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java]: Let's undo the 3 whitespace changes below

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/SetSchemaMetadata.java [hunk 14]
```diff
@@ -39,18 +39,22 @@ public abstract class SetSchemaMetadata<R extends ConnectRecord<R>> implements T
     public static final String OVERVIEW_DOC =
             "Set the schema name, version or both on the record's key (<code>" + Key.class.getName() + "</code>)"
                     + " or value (<code>" + Value.class.getName() + "</code>) schema.";
-
```
Comment (mimaison) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/SetSchemaMetadata.java]: Let's keep this empty line that separates the fields from the interface but remove the new empty line added just below.

File: connect/transforms/src/test/java/org/apache/kafka/connect/transforms/SetSchemaMetadataTest.java [hunk 15]
```diff
@@ -104,6 +115,38 @@ public void schemaNameAndVersionUpdateWithStruct() {
...
+        props.put("schema.version", "42");
+        props.put("replace.null.with.default", replaceNullWithDefault);
+        xform.configure(props);
+
+        final SinkRecord record = new SinkRecord("", 0, null, null, schema, value, 0);
+
+        final SinkRecord updatedRecord = xform.apply(record);
+
+        assertEquals("foo", updatedRecord.valueSchema().name());
+        assertEquals(Integer.valueOf(42), updatedRecord.valueSchema().version());
```
Comment (mimaison) [inline connect/transforms/src/test/java/org/apache/kafka/connect/transforms/SetSchemaMetadataTest.java]: We can just use `42`, no need to explicitly get an `Integer`

File: connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java [hunk 16]
```diff
@@ -186,7 +188,94 @@ public void nonExistentNestedFieldWithSchemaShouldFail() {
 
     @Test
     public void testExtractFieldVersionRetrievedFromAppInfoParser() {
-        assertEquals(AppInfoParser.getVersion(), xform.version());
+        assertEquals(AppInfoParser.getVersion(), xformKey.version());
+    }
+
+    @Test
+    public void testUnsetOptionalFieldIsNull() {
```
Comment (mimaison) [inline connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java]: Is there a reason why you used `ParameterizedTest` in some places and multiple tests in others?

**Review by mimaison — COMMENTED** [review]
Thanks for the PR! I made another pass. The changes look fine overall, I left a couple of questions/suggestions and spotted a bunch of cosmetic issues.

File: connect/transforms/src/test/java/org/apache/kafka/connect/transforms/SetSchemaMetadataTest.java (hunk 7 above)
Comment (mfvitale) [reply connect/transforms/src/test/java/org/apache/kafka/connect/transforms/SetSchemaMetadataTest.java:126]: I would like to differentiate with the `expectedValue` variable that can have different value based on the test.

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java (hunk 11 above)
Comment (mfvitale) [reply connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java]: I used the same name used in `JsonConverter`  to be consistent but I'm fine with the renaming.

File: connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java (hunk 16 above)
Comment (mfvitale) [reply connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java]: Just changed to be a `ParameterizedTest`

**Comment by mfvitale** [comment]
> Thanks for the PR! I made another pass. The changes look fine overall, I left a couple of questions/suggestions and spotted a bunch of cosmetic issues.

Changes done.

**Review by mimaison — APPROVED** [review]
Thanks for the PR! LGTM

**Comment by mimaison** [comment]
@C0urante @yashmayya @gharris1727 Since you also voted on the KIP, do you want to take a look?
If not let's merge next week. Thanks

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java [hunk 17]
```diff
@@ -67,6 +67,7 @@ public abstract class Cast<R extends ConnectRecord<R>> implements Transformation
                     + "or value (<code>" + Value.class.getName() + "</code>).";
 
     public static final String SPEC_CONFIG = "spec";
+    public static final String REPLACE_NULL_WITH_DEFAULT = "replace.null.with.default";
```
Comment (gharris1727) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java]: Config key constants are generally suffixed with `_CONFIG`
```suggestion
    public static final String REPLACE_NULL_WITH_DEFAULT_CONFIG = "replace.null.with.default";
```

Here and throughout.

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/HeaderFrom.java [hunk 18]
```diff
@@ -97,6 +100,7 @@ static Operation fromName(String name) {
         public String toString() {
             return name;
         }
+
```
Comment (gharris1727) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/HeaderFrom.java]: nit: revert this spacing change

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java [hunk 19]
```diff
@@ -181,7 +186,7 @@ public Object valueFrom(Struct struct) {
         }
 
         if (current.schema().field(lastStep()) != null) {
-            return current.get(lastStep());
+            return withDefault ? current.get(lastStep()) : current.getWithoutDefault(lastStep());
```
Comment (gharris1727) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java]: nit: call lastStep once.

**Review by gharris1727 — COMMENTED** [review]
I had a few small comments, but nothing serious. The functionality itself looks good. Thanks @mfvitale for the feature!

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java (hunk 17 above)
Comment (mfvitale) [reply connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java]: This was the original name but @mimaison asked to remove the suffix to be aligned with others fields. https://github.com/apache/kafka/pull/15756#discussion_r1662794000

Comment (mimaison) [reply connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java]: In some files we have the `_CONFIG` suffix, for example `Cast` but in some others like `HeadersFrom` we don't. When I asked you to make changes I meant so it's consistent in each file. Sorry if I wasn't clear enough.

Comment (mfvitale) [reply connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java]: So you meant consistent per file not at whole. It seems so weird but ok I'll then adjust it to be per file consistent.

Comment (mfvitale) [reply connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java]: @gharris1727 Changes done.

Comment (gharris1727) [reply connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Cast.java]: Sorry I didn't look at the previous review comments to see that this was already discussed. I am fine with _CONFIG everywhere, or _CONFIG only in files that have _CONFIG for the other variables. Thank you for changing this to be more consistent!

**Review by gharris1727 — APPROVED** [review]
LGTM, just wanted to note that there are still two transformations remaining, ValueToKey and ReplaceField

**Comment by gharris1727** [comment]
Test failures appear unrelated, and connect:transforms:test passes for me locally.

**Comment by gharris1727** [comment]
Thank you @mfvitale for driving the KIP and implementation!
